src/disp_s1/solid_earth_tides.py

```diff
@@ -0,0 +1,247 @@
+    atr: dict,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Generate time ranges for reference and secondary files."""
+    ref_time_range = pd.to_datetime(
+        np.linspace(
+            reference_start_time.value, reference_stop_time.value, atr["LENGTH"]
+        )
+    )
+    sec_time_range = pd.to_datetime(
+        np.linspace(
+            secondary_start_time.value, secondary_stop_time.value, atr["LENGTH"]
+        )
+    )
+    ref_time_tile = np.tile(ref_time_range, (atr["WIDTH"], 1)).T
+    sec_time_tile = np.tile(sec_time_range, (atr["WIDTH"], 1)).T
+    return ref_time_tile, sec_time_tile
+
+
+def calculate_solid_earth_tides_correction(
+    ifgram_filename: Filename,
+    reference_cslc_file: Filename,
+    secondary_cslc_file: Filename,
+    los_east_file: Filename,
+    los_north_file: Filename,
+    reference_point: Optional[ReferencePoint] = None,
+) -> np.ndarray:
+    """Calculate the relative displacement correction for solid earth tides."""
+    # Load bounds, transform, CRS, and shape from the unwrapped interferogram
+    bounds, crs, affine_transform, (height, width) = load_raster_bounds_and_transform(
+        ifgram_filename
+    )
+    bounds_geo = transform_bounds_to_epsg4326(bounds, crs)
+    atr = generate_atr(bounds_geo, height, width)
+
+    # Extract timing information from the CSLC files
+    reference_start_time = pd.to_datetime(
+        get_zero_doppler_time(reference_cslc_file, type_="start")
+    )
+    reference_stop_time = pd.to_datetime(
+        get_zero_doppler_time(reference_cslc_file, type_="end")
+    )
+    secondary_start_time = pd.to_datetime(
+        get_zero_doppler_time(secondary_cslc_file, type_="start")
+    )
+    secondary_stop_time = pd.to_datetime(
+        get_zero_doppler_time(secondary_cslc_file, type_="end")
+    )
+
+    # Generate the lat/lon arrays for the SET geogrid
+    lat_geo_array = np.linspace(
+        atr["Y_FIRST"],
+        atr["Y_FIRST"] + atr["Y_STEP"] * atr["LENGTH"],
+        num=atr["LENGTH"],
+    )
+    lon_geo_array = np.linspace(
+        atr["X_FIRST"], atr["X_FIRST"] + atr["X_STEP"] * atr["WIDTH"], num=atr["WIDTH"]
+    )
+    lat_geo_mesh, lon_geo_mesh = np.meshgrid(
+        lat_geo_array, lon_geo_array, indexing="ij"
+    )
+
+    # Generate time ranges
+    ref_time_tile, sec_time_tile = calculate_time_ranges(
+        reference_start_time,
+        reference_stop_time,
+        secondary_start_time,
+        secondary_stop_time,
+        atr,
+    )
+
+    # Prepare resolution arrays
+    res_lat_arr = np.ones(ref_time_tile.shape) * atr["Y_STEP"]
+    res_lon_arr = np.ones(ref_time_tile.shape) * atr["X_STEP"]
+
+    # Parallelize grid calculation using ThreadPoolExecutor
+    with ThreadPoolExecutor() as executor:
+        ref_input_data = zip(
+            ref_time_tile.ravel(),
+            lat_geo_mesh.ravel(),
+            lon_geo_mesh.ravel(),
+            res_lat_arr.ravel(),
+            res_lon_arr.ravel(),
+        )
```

Comment:
Just noting here in case we test and this uses a too much memory- another option here instead of building the full 2D array and `.ravel()`-int is to use [`itertools.product`](https://docs.python.org/3/library/itertools.html#itertools.product)